src/plugins/intel_gpu/src/graph/include/reshape_inst.h

```diff
@@ -23,10 +23,69 @@ struct typed_program_node<reshape> : public typed_program_node_base<reshape> {
 
     program_node& input() const { return get_dependency(0); }
 
+    bool has_padding() const {
+        return (this->get_output_layout().data_padding || input().get_output_layout(false).data_padding);
+    }
+
+    bool has_outer_padding_offset() const {
+        if (!has_padding())
+            return false;
+
+        auto input_layout = input().get_output_layout(false);
+        auto input_pad = input_layout.data_padding;
+        bool spatial_pad = false;
+        for (size_t i = 0 ; i < input_layout.get_spatial_rank() ; i++)
+            spatial_pad |= (input_pad.lower_size().spatial[i] != 0);
+
+        if (input_pad.lower_size().feature[0] != 0 || spatial_pad)
+            return false;
+
+        // Outer padding exists. It might need to update padding size of output layout
+        return true;
+    }
+
     bool is_in_place() const {
         if (this->is_output() || this->has_fused_primitives())
             return false;
-        return (!this->get_output_layout().data_padding && !input().get_output_layout(false).data_padding);
+
+        if (has_padding()) {
+            if (!this->is_dynamic() && has_outer_padding_offset() &&
+                this->get_users().front()->get_preferred_impl_type() == impl_types::onednn)
+                return true;
+            else
+                return false;
+        }
+
+        return true;
+    }
+
+    void adjust_output_padding() {
+        if (!is_in_place() || !has_padding())
+            return;
+
+        auto input_layout = input().get_output_layout(false);
+        auto output_layout = this->get_output_layout();
+        if (input_layout.batch() != output_layout.batch()) {
+            auto input_pad = input_layout.data_padding;
+            size_t first_element_offset = input_pad.lower_size().batch[0];
+            for (size_t i = 0 ; i < input_layout.get_spatial_rank() ; i++) {
+                first_element_offset *= input_layout.spatial(i);
+            }
+            first_element_offset *= input_layout.feature();
+
+            size_t inner_size = 1;
+            for (size_t i = 0 ; i < output_layout.get_spatial_rank() ; i++) {
+                inner_size *= output_layout.spatial(i);
+            }
+            inner_size *= output_layout.feature();
```

Comment:
this code assumes that the reshape may change batch size as well. But do we have such case in swin/twin/vit? If not, I think we can simplify the code here..

Reply:
Yes, there is this case in swin.